### PR TITLE
fix: allow elevator misalignment

### PR DESCRIPTION
**Pull request by scarf005**
#### Summary

SUMMARY: Bugfixes "Fix misaligned elevator not working due to misalignment"

#### Purpose of change

elevator misalignment was making elevator control not working.

#### Describe the solution

1. aligned sarcophagus elevator.
2. made elevator checks more forgiving (now considers up to 3-tile radius)
~~3. inlined functions 'cause godbolt results told me they were more efficient~~

~~[no inline][no-inline] (133 loc) vs [inline][inline] (99 loc)~~

[no-inline]: https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:8,positionColumn:1,positionLineNumber:8,selectionStartColumn:1,selectionStartLineNumber:8,startColumn:1,startLineNumber:8),source:'%23include+%3Ciostream%3E%0A%23include+%3Cvector%3E%0A%0Ausing+namespace+std%3B%0A%0Aauto+xs+%3D+vector%7B1,2,3,4%7D%3B%0A%0Aauto+f()+-%3E+void+%7B%0A++++auto%26+a+%3D+xs%3B%0A++++a.push_back(5)%3B%0A%7D%0A%0Aauto+g()+-%3E+void+%7B%0A++++auto%26+a+%3D+xs%3B%0A++++a.push_back(6)%3B%0A%7D%0A%0Aauto+main()+-%3E+int+%7B%0A++++f()%3B%0A++++g()%3B%0A%0A++++for+(+const+auto%26+x+:+xs+)+%7B%0A++++++++cout+%3C%3C+x+%3C%3C+%22,+%22%3B%0A++++%7D%0A++++cout+%3C%3C+!'%5Cn!'%3B%0A++++return+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:43.43509859858206,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g132,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B17+-O2',overrides:!((name:edition,value:'2021')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+13.2+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'rustc+nightly',editorid:1,fontScale:16,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+13.2+(Compiler+%231)',t:'0')),k:56.56490140141794,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

[inline]: https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:27,positionColumn:2,positionLineNumber:27,selectionStartColumn:2,selectionStartLineNumber:27,startColumn:2,startLineNumber:27),source:'%23include+%3Ciostream%3E%0A%23include+%3Cvector%3E%0A%0Ausing+namespace+std%3B%0A%0Aauto+xs+%3D+vector%7B1,2,3,4%7D%3B%0A%0Ainline+auto+f()+-%3E+void+%7B%0A++++auto%26+a+%3D+xs%3B%0A++++a.push_back(5)%3B%0A%7D%0A%0Ainline+auto+g()+-%3E+void+%7B%0A++++auto%26+a+%3D+xs%3B%0A++++a.push_back(6)%3B%0A%7D%0A%0Aauto+main()+-%3E+int+%7B%0A++++f()%3B%0A++++g()%3B%0A%0A++++for+(+const+auto%26+x+:+xs+)+%7B%0A++++++++cout+%3C%3C+x+%3C%3C+%22,+%22%3B%0A++++%7D%0A++++cout+%3C%3C+!'%5Cn!'%3B%0A++++return+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:43.43509859858206,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g132,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B17+-O2',overrides:!((name:edition,value:'2021')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+13.2+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'rustc+nightly',editorid:1,fontScale:16,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+13.2+(Compiler+%231)',t:'0')),k:56.56490140141794,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

#### Tests

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/1166a81e-fb2d-4284-8d52-f2fb0971b09a

### CODE REVIEW

**Comment by olanti-p** [comment]
`inline` is best suited for hot, or small and simple (like in your examples) functions where the call cost is noticeable relative to the work done. Using `inline` on large functions will inflate the size of the binary, which can actually worsen the performance on some CPUs, or it will simply do nothing because `inline` does not _force_ the compiler to inline, it just advises it to, and the compiler is free to choose whether to inline the function or not regardless of whether the function has `inline` or not. It is hard to tell without a profiler whether you _need_ to inline, and it can be hard to tell without disassembling the resulting binary whether the inline even worked.

In your 133 loc example, from what I can tell from the phone the compiler _does_ inline the functions. It just does not eliminate functions themselves because they're not `static`, so the compiler thinks they can be used from other translation units, which would be impossible if they were eliminated.

TLDR: It is ok to have call overhead if you don't have a specific reason to avoid it.

**Comment by chaosvolt** [comment]
```
../tests/translations_test.cpp:147: warning:
  Skipped (unable to set locale)

D:\a\_temp\bfbeaa38-6fcb-4d6b-806d-d1350dacae97: line 1:  1382 Segmentation fault      ./tests/cata_test.exe --rng-seed time
```
Unrelated I hope?

File: src/iexamine_elevator.cpp [hunk 1]
```diff
@@ -16,6 +17,16 @@
...
+/// poor person's https://en.cppreference.com/w/cpp/utility/optional/and_then
+template <typename T, typename Fn>
+auto and_then( std::optional<T> const &opt, Fn &&f ) -> std::optional<std::invoke_result_t<Fn, T>>
+{
+    if( opt ) {
+        return std::optional{f( *opt )};
+    }
+    return std::nullopt;
+}
+
```
Comment (olanti-p) [inline src/iexamine_elevator.cpp:29]: Maybe move it to some header?
`cata_utility.h`, or some other.

Comment (scarf005) [reply src/iexamine_elevator.cpp:29]: moved to `cata_algo.h`, as it's only referenced by 3 src files wheras `cata_utility.h` is imported from 109 files

**Comment by scarf005** [comment]
@chaosvolt i think this one is ready too

**Review by chaosvolt — APPROVED** [review]
